src/ir_passes/lower/build_ctx.rs

```diff
@@ -0,0 +1,347 @@
+use std::{collections::HashMap, iter, rc::Rc};
+
+use crate::ir::{self, Ctx};
+use calyx::structure;
+use calyx_ir::{self as calyx, RRC};
+use itertools::Itertools;
+
+use super::utils as compile_utils;
+use super::Fsm;
+
+/// Bindings associated with the current compilation context
+#[derive(Default)]
+pub(super) struct Binding {
+    // Component signatures
+    comps: HashMap<ir::CompIdx, RRC<calyx::Cell>>,
+    /// Mapping to the component representing FSM with particular number of states
+    pub fsm_comps: HashMap<u64, calyx::Component>,
+}
+
+impl Binding {
+    /// Inserts a [calyx::Cell] into the binding
+    pub fn insert(&mut self, name: ir::CompIdx, sig: RRC<calyx::Cell>) {
+        self.comps.insert(name, sig);
+    }
+
+    /// Gets a [calyx::Cell]'s signature from an [ir::CompIdx]
+    pub fn get(&self, idx: &ir::CompIdx) -> Option<Vec<calyx::PortDef<u64>>> {
+        self.comps.get(idx).map(Self::cell_to_port_def)
+    }
+
+    /// Converts a cell to a list of port definitions
+    pub fn cell_to_port_def(cr: &RRC<calyx::Cell>) -> Vec<calyx::PortDef<u64>> {
+        let cell = cr.borrow();
+        cell.ports()
+            .iter()
+            .map(|pr| {
+                let port = pr.borrow();
+                // Reverse port direction because signature refers to internal interface.
+                (port.name, port.width, port.direction.reverse()).into()
+            })
+            .collect()
+    }
+}
+
+/// Contains the context needed to compile a component.
+pub(super) struct BuildCtx<'a> {
+    pub builder: calyx::Builder<'a>,
+    pub binding: &'a mut Binding,
+    pub comp: &'a ir::Component,
+    ctx: &'a ir::Context,
+    lib: &'a calyx::LibrarySignatures,
+    fsms: HashMap<ir::EventIdx, Fsm>,
+    /// Mapping from [ir::InstIdx]s to the calyx cell instantiated.
+    instances: HashMap<ir::InstIdx, RRC<calyx::Cell>>,
+    /// Mapping from [ir::InstIdx]s to a reference of the calyx cell instantiated/invoked
+    invokes: HashMap<ir::InvIdx, RRC<calyx::Cell>>,
```

Comment:
If all instances and all invokes will have a binding here, we should instead use `DenseIndexInfo` instead

Reply:
Instances and invokes that are deleted will not be in this mapping.